File: backend-app/src/main/java/tom/projects/monkeylog/controller/ExerciseController.java
package tom.projects.monkeylog.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;
import tom.projects.monkeylog.dto.exercise.ExerciseCreateRequest;
import tom.projects.monkeylog.dto.exercise.ExerciseResponse;
import tom.projects.monkeylog.dto.exercise.ExerciseTypeResponse;
import tom.projects.monkeylog.dto.exercise.ExerciseUpdateRequest;
import tom.projects.monkeylog.mapper.ExerciseMapper;
import tom.projects.monkeylog.service.ExerciseService;

import java.util.List;

@RestController
@RequiredArgsConstructor
public class ExerciseController {
    private final ExerciseService exerciseService;
    private final ExerciseMapper exerciseMapper;

    @GetMapping("/exercises")
    List<ExerciseResponse> all() {
        return exerciseMapper.exercisesToExerciseResponses(exerciseService.all());
    }

    @GetMapping("/exercises/{id}")
    ExerciseResponse get(@PathVariable Long id) {
        return exerciseMapper.exerciseToExerciseResponse(exerciseService.get(id));
    }

    @PostMapping("/exercises")
    ExerciseResponse save(@RequestBody ExerciseCreateRequest exerciseCreateRequest) {
        return exerciseMapper.exerciseToExerciseResponse(exerciseService.save(exerciseCreateRequest));
    }

    @PatchMapping("/exercises/{id}")
    ExerciseResponse update(@RequestBody ExerciseUpdateRequest exerciseUpdateRequest, @PathVariable Long id) {
        return exerciseMapper.exerciseToExerciseResponse(exerciseService.update(exerciseUpdateRequest, id));
    }

    @DeleteMapping("/exercises/{id}")
    void destroy(@PathVariable Long id) {
        exerciseService.delete(id);
    }

    @GetMapping("/exercise-types")
    List<ExerciseTypeResponse> getExerciseTypes() {
        return exerciseMapper.exerciseTypesToExerciseTypeResponses(exerciseService.getExerciseTypes());
    }
}

File: backend/src/main/kotlin/tom/com/monkeylog/controller/ExerciseController.kt
package tom.com.monkeylog.controller

import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import tom.com.monkeylog.dto.workout.ExerciseCategoryResponse
import tom.com.monkeylog.dto.workout.ExerciseCreateRequest
import tom.com.monkeylog.dto.workout.ExerciseUpdateRequest
import tom.com.monkeylog.mapper.toResponse
import tom.com.monkeylog.model.exercise.ExerciseCategory
import tom.com.monkeylog.service.ExerciseService
import java.util.*

@RestController
@Tag(name = "Exercise")
class ExerciseController(
    private val exerciseService: ExerciseService,
) {
    @GetMapping("/exercises")
    fun getExercises() = exerciseService.allExercises().map { it.toResponse() }

    @GetMapping("/exercises/{id}")
    fun getExercise(@PathVariable id: UUID) = exerciseService.getExercise(id).toResponse()

    @PostMapping("/exercises")
    fun createExercise(@RequestBody exerciseCreateRequest: ExerciseCreateRequest) =
        exerciseService.saveExercise(exerciseCreateRequest).toResponse()

    @PatchMapping("/exercises/{id}")
    fun updateExercise(
        @RequestBody exerciseUpdateRequest: ExerciseUpdateRequest,
        @PathVariable id: UUID
    ) = exerciseService.updateExercise(exerciseUpdateRequest, id).toResponse()

    @DeleteMapping("/exercises/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteExercise(@PathVariable id: UUID) = exerciseService.deleteExercise(id)

    @GetMapping("/exercises/categories")
    fun getExerciseCategories(): Map<ExerciseCategory, ExerciseCategoryResponse> {
        return ExerciseCategory.values().associateWith { ExerciseCategoryResponse(it.statistics, it.validFields) }
    }
}